Assets/Mirror/Runtime/Compression.cs

```diff
@@ -0,0 +1,247 @@
+                    c = w;
+                    break;
+                case 3:
+                    a = x;
+                    b = y;
+                    c = z;
+                    break;
+                default:
+                    return Vector3.zero;
+            }
+
+            return new Vector3(a, b, c);
+        }
+
+
+        /// <summary>
+        /// Used to read a Compressed Quaternion from 4 bytes
+        /// <para>Quaternion is normalized</para>
+        /// </summary>
+        public static Quaternion DecompressQuaternion(uint packed)
+        {
+            // 10 bits
+            const uint mask = 0b11_1111_1111;
+            Quaternion result;
+
+
+            uint a = packed & mask;
+            uint b = (packed >> 10) & mask;
+            uint c = (packed >> 20) & mask;
+            uint largestIndex = (packed >> 30) & mask;
+
+            float x = ScaleFromUInt(a, QuaternionMinValue, QuaternionMaxValue, QuaternionUintRange);
+            float y = ScaleFromUInt(b, QuaternionMinValue, QuaternionMaxValue, QuaternionUintRange);
+            float z = ScaleFromUInt(c, QuaternionMinValue, QuaternionMaxValue, QuaternionUintRange);
+
+            Vector3 small = new Vector3(x, y, z);
+            result = FromSmallerDimensions(largestIndex, small);
+            return result;
+        }
+
+        static Quaternion FromSmallerDimensions(uint largestIndex, Vector3 smallest)
+        {
+            float a = smallest.x;
+            float b = smallest.y;
+            float c = smallest.z;
+
+            float x, y, z, w;
+            float largest = Mathf.Sqrt(1 - a * a - b * b - c * c);
+            switch (largestIndex)
+            {
+                case 0:
+                    x = largest;
+                    y = a;
+                    z = b;
+                    w = c;
+                    break;
+                case 1:
+                    x = a;
+                    y = largest;
+                    z = b;
+                    w = c;
+                    break;
+                case 2:
+                    x = a;
+                    y = b;
+                    z = largest;
+                    w = c;
+                    break;
+                case 3:
+                    x = a;
+                    y = b;
+                    z = c;
+                    w = largest;
+                    break;
+                default:
+                    return Quaternion.identity;
+            }
+            // 0.999999f is tolerance for Unity's Dot function
+            Debug.Assert(x * x + y * y + z * z + w * w < (2 - 0.999999f), "larger than 1");
+            return new Quaternion(x, y, z, w).normalized;
+        }
+
+
+        /// <summary>
+        /// Scales float from minFloat->maxFloat to 0->maxUint
+        /// <para>values out side of minFloat/maxFloat will return either 0 or maxUint</para>
+        /// </summary>
+        public static uint ScaleToUInt(float value, float minFloat, float maxFloat, uint maxUint)
```

Comment:
what about minUint?
if you add that part then this will be a fully fledged 'scale float to uint' function that we will be able to use from other classes too :)

Reply:
I think adding `minUint` would overcomplicate this, the idea would be to scale a float to `0->maxUint` then use bitwise afterwards. It might even be worth changing `uint maxUint` back to bit count so it doesn't have to be manually calculated each time.

Is there a reason I am missing that would cause someone to need the min to not be 0?

Reply:
scaling to smaller data types can be extremely useful for compression.
might as well have the complete function so we never have to worry about it again :)

maybe some day someone wants to scale his first value to 0..127 and his second value to 128..255 or similar.